engine/grid.py

```diff
@@ -300,73 +307,164 @@ def find_border_nodes(self):
             'border_nodes' will be initialized.
         """
         border_list = []
-        cols = self.nodes.shape[1]
-        rows = self.nodes.shape[0]
         leftmost_node = None
+        rightmost_node = None
         leftmost_node_pos = None
-        for row in range(rows):
-            for col in range(cols):
+        rightmost_node_pos = None
+        
+        for row in range(self.num_rows):
+            for col in range(self.num_cols):
                 node = self.nodes[row][col]
-                if node.is_active and self.is_on_border(row, col, rows-1, cols-1):
+                #border node 1 left active node 
+                if node.is_active and self.is_on_border(row, col, self.num_rows-1, self.num_cols-1):
                     # check if this is an active node and on the border
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
                         leftmost_node = node
                         leftmost_node_pos = (row, col)
+                if node.is_active and self.is_on_border(row, col, self.num_rows , self.num_cols):
+                    # check if this is an active node and on the border
+                    self.nodes[row][col].is_border = True
+                    if rightmost_node_pos is None or row > rightmost_node_pos[0]:
+                        rightmost_node = node
+                        rightmost_node_pos = (row, col)
+                    
+                    
+                        
         self.border_nodes = border_list
         self.leftmost_node = leftmost_node
         self.leftmost_node_pos = leftmost_node_pos
+        self.rightmost_node = rightmost_node
+        self.rightmost_node_pos = rightmost_node_pos
 
-    # Return bottom most node that is activated in the right column
-
+    
+    ##Return bottom most node that is activated in the right column
     def bottom_rightmost_node(self, pos):
-
         # node_info: (node, row, col)
         candidate_nodes = [node_info for node_info in self.border_nodes if node_info[2] == pos[1]+1]
         if (candidate_nodes == []):
             return None
         else:
             node_info = min(candidate_nodes,key=lambda node_info: node_info[1])
             return (node_info[1],node_info[2])
+    
+    def bottom_leftmost_node(self, pos):
```

Comment:
is there/do we need tests?

Reply:
I don't think we have unit test for anything. The only check is the tests.functionality_tests.test_activation_shapes  we have been doing which draws the graphs

Reply:
It will be a great future step to do more tests because there are probably more bugs